File: 198-house-robber/house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums:
            return 0
        
        n = len(nums)
        if n == 1:
            return nums[0]

        prev_house = nums[0]
        curr_house = max(nums[0], nums[1])
        max_profit = 0

        for i in range(2, n):
            max_profit = max(curr_house, nums[i] + prev_house)
            prev_house = curr_house
            curr_house = max_profit

        return curr_house

File: 198-house-robber/house-robber.ts
function rob(nums: number[]): number {
    const dp: number[] = []
    dp[0] = nums[0]
    dp[1] = Math.max(nums[0], nums[1])

    for (let i = 2; i < nums.length; i++) {
        dp[i] = Math.max(dp[i - 1], nums[i] + dp[i - 2])
    }
    return dp[nums.length-1]
};